### PR TITLE
afl++ 4.21c (new formula)

### PR DESCRIPTION
**Pull request by henryhchchc**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

**Comment by henryhchchc** [comment]
```
afl++
    * Binaries built for a non-native architecture were installed into afl++'s prefix.
      The offending files are:
        /home/linuxbrew/.linuxbrew/Cellar/afl++/4.21/share/afl/testcases/others/elf/small_exec.elf	(i386)
  Error: 1 problem in 1 formula detected.
```

I think this checking can be bypassed since this file is the data used for seeding fuzzing inputs. It is not intended to be executed.

**Comment by henryhchchc** [comment]
```
Indirect dependencies with linkage:
    python@3.12
```
And, what does this mean?

File: Formula/a/afl++.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
+class Aflxx < Formula
+  desc "American Fuzzy Lop++"
+  homepage "https://aflplus.plus"
+  url "https://github.com/AFLplusplus/AFLplusplus/archive/refs/tags/v4.21c.tar.gz"
+  sha256 "11f7c77d37cff6e7f65ac7cc55bab7901e0c6208e845a38764394d04ed567b30"
+  license "Apache-2.0"
+
+  depends_on "cmake" => :build
+  depends_on "coreutils" => :build
+  depends_on "make" => :build
```
Comment (carlocab) [inline Formula/a/afl++.rb:10]: Can't we rely on system `make`?

File: Formula/a/afl++.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  depends_on "wget" => :build
```
Comment (carlocab) [inline Formula/a/afl++.rb:11]: Why do we need this? Formulae should not be downloading anything at build-time.

**Review by carlocab — COMMENTED** [review]
Thanks for the PR. A few questions/comments.

File: audit_exceptions/mismatched_binary_allowlist.json [hunk 3]
```diff
@@ -1,4 +1,5 @@
 {
+  "afl++": "share/afl/testcases/others/elf/small_exec.elf",
```
[the remark this answers is not in the document]
Comment (henryhchchc) [reply audit_exceptions/mismatched_binary_allowlist.json:2]: AFL++ maintains a set of common file formats (e.g., text, images) to seed test inputs. It happens that one of these files are in ELF.

File: Formula/a/afl++.rb (hunk 1 above)
Comment (henryhchchc) [reply Formula/a/afl++.rb]: Humm I grabbed the dependencies from their build instructions. Seems the build works without them so they are removed.

File: Formula/a/afl++.rb [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  depends_on "llvm"
+  depends_on "python@3.12"
+
+  def install
+    ENV.prepend_path "PATH", Formula["coreutils"].libexec/"gnubin"
+
+    ENV["CC"] = Formula["llvm"].opt_bin/"clang"
+    ENV["CXX"] = Formula["llvm"].opt_bin/"clang++"
```
[the remark this answers is not in the document]
Comment (henryhchchc) [reply Formula/a/afl++.rb]: Their build instructions suggest the use of clang. Seems it also works with gcc. These two lines are therefore removed.

File: Formula/a/afl++.rb [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+    inreplace "GNUmakefile.llvm", "-flat_namespace", ""
```
[the remark this answers is not in the document]
Comment (carlocab) [reply Formula/a/afl++.rb:15]: Given the comments around it, I doubt they really know what it's for either. They need to get rid of `-undefined suppress` too and replace it with `-undefined dynamic_lookup`.

**Review by carlocab — COMMENTED** [review]
Could you squash all your commits together into a single one? Thanks!

**Comment by henryhchchc** [comment]
> Could you squash all your commits together into a single one? Thanks!

Done.

File: Formula/a/afl++.rb [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+  license "Apache-2.0"
+
+  depends_on "coreutils" => :build
+  depends_on "llvm"
+  depends_on "python@3.12"
+
+  def install
+    ENV.prepend_path "PATH", Formula["coreutils"].libexec/"gnubin"
+
+    inreplace "GNUmakefile.llvm", "-flat_namespace", ""
```
Comment (carlocab) [inline Formula/a/afl++.rb:15]: Let's fix this properly:
```suggestion
    inreplace "GNUmakefile.llvm" do |s|
      s.gsub! "-Wl,-flat_namespace", ""
      s.gsub! "-undefined,suppress", "-undefined,dynamic_lookup"
    end
```

**Review by carlocab — COMMENTED** [review]
One suggestion, and then I'm fine with this.

**Review by carlocab — APPROVED** [review]
Thanks!

**Pull request merged**